### DISCUSSION TITLE
catching error with different observable without completing

### DISCUSSION THREAD

**Discussion by JeromeDeLeon** [ReactiveX/rxjs#6927] [Q&A] [answered]
Let's say I have subject that emits incremental value and throws an error at some point but I still want to continue the emission say when it catches an error, instead of retrying or rethrowing, I pass default value.

Here's the [code sample](https://stackblitz.com/edit/o3ya3a?devtoolsheight=50).

Passing different observable automatically completes the observable which is not I want to achieve.
I thought returning any result from the the `selector` would continue the chain.
Is there any other way I would achieve this than `catchError((err, caught) => caught.pipe(startWith(4)))` or is it because to make sure we're able to trigger `finalize` after we catch an error?

**Comment by josepot** [comment, marked as answer]
Hi @JeromeDeLeon !

Thanks for sharing a code example, that helped me understand your questions.

> Passing different observable automatically completes the observable which is not I want to achieve.

That's because the observable to that you were returning inside the selector function of `catchError` (`of(4)`) is an Observable that emits and then completes. Meaning that `of(4)` is the same as:

```ts
new Observable<number>(observer => {
  observer.next(4);
  observer.complete();
});
```

> I thought returning any result from the the selector would continue the chain.

Not exactly, it will continue with the provided Observable. If the provided Observable completes, then the resulting Observable will complete, if it doesn't, then the returning Observable won't complete.

> Is there any other way I would achieve this than catchError((err, caught) => caught.pipe(startWith(4)))

Sure, there are many ways to accomplish the same thing. For instance:

```ts
catchError((err, caught) => concat(of(4), caught))
```

> or is it because to make sure we're able to trigger finalize after we catch an error?

Apologies, but I don't quite understand what you mean here.

**Reply by JeromeDeLeon** [reply]
Your first statement already answered my last question. Thank you. I'll try your suggestion and see how it goes.

**Reply by voliva** [reply]
Just to add a bit more info, when an Observable emits an error, it's completely dead: It can't emit more values. The only way is to resubscribe to it, but it will start from the beginning.

This means that this observable:
```ts
new Observable<number>(observer => {
  observer.next(1);
  observer.next(2);
  observer.next(3);
  observer.error("woops");
  observer.next(4);
});
```

Will never get to emit the `4` value. When it emits the error, it's cleaned up. If you resubscribe then it will start back at `1 - 2 - 3` then error again.

`catchError` only stops the propagation of that error further down and lets you replace it with another observable. It can be the same observable, in which case it will just resubscribe and start from the beginning.

**Reply by JeromeDeLeon** [reply]
Sorry for taking this so long, but I think I get it working on my side. As I was struggling up until now, all the replies above are correct and helped me more to understand rxjs.

What I ended up doing was to create an inner observable and pipe the error there. Pretty much similar to what [Ben Lesh talk about](https://youtu.be/3LKMwkuK0ZE?t=20m17s)
